src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+
+namespace System.Net.ServerSentEvents
+{
+    internal sealed class PooledByteBufferWriter : IBufferWriter<byte>, IDisposable
+    {
+        private byte[] _rentedBuffer;
+        private int _index;
+        private const int MinimumBufferSize = 256;
+
+        // Value copied from Array.MaxLength in System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Array.cs.
+        public const int MaximumBufferSize = 0X7FFFFFC7;
+
+        public PooledByteBufferWriter(int initialCapacity = MinimumBufferSize)
+        {
+            Debug.Assert(initialCapacity > 0);
+
+            _rentedBuffer = ArrayPool<byte>.Shared.Rent(initialCapacity);
+            _index = 0;
+        }
+
+        public ReadOnlyMemory<byte> WrittenMemory
+        {
+            get
+            {
+                Debug.Assert(_rentedBuffer != null);
+                Debug.Assert(_index <= _rentedBuffer.Length);
+                return _rentedBuffer.AsMemory(0, _index);
+            }
+        }
+
+        public int WrittenCount
+        {
+            get
+            {
+                Debug.Assert(_rentedBuffer != null);
+                return _index;
+            }
+        }
+
+        public int Capacity
+        {
+            get
+            {
+                Debug.Assert(_rentedBuffer != null);
+                return _rentedBuffer.Length;
+            }
+        }
+
+        public void Reset()
+        {
+            Debug.Assert(_rentedBuffer != null);
+            Debug.Assert(_index <= _rentedBuffer.Length);
+
+            _rentedBuffer.AsSpan(0, _index).Clear();
```

Comment:
Why is this needed?

Reply:
It's following practice used in STJ and specifically the PooledByteBufferWriter that it got copied from.

Reply:
Is it clearing buffers before returning them not considered good practice?

Reply:
> Is it clearing buffers before returning them not considered good practice?

a) This isn't returning it. It's resetting it for the next write.
b) We generally don't on return except for specific circumstances, like being in a library like crypto that wants extra scrutiny or returning arrays of reference types to avoid rooting objects that may be stored in the array. The default for Return is to not clear. For example, the WriteAsync polyfill earlier is not clearing.

Reply:
I've refactored the code following Miha's suggestion, but it seems STJ is susceptible to the same issue. Effectively, the buffer is being cleared every time we flush to the underlying stream.

https://github.com/dotnet/runtime/blob/d4516bf13c0a06d955912c15e71a13e3dc894db0/src/libraries/Common/src/System/Text/Json/PooledByteBufferWriter.cs#L108

I'll file an issue.